cassovary-core/src/test/scala/com/twitter/cassovary/graph/ConcurrentHashMapDynamicGraphSpec.scala

```diff
@@ -0,0 +1,77 @@
+package com.twitter.cassovary.graph
+
+import org.scalatest.{Matchers, WordSpec}
+
+class ConcurrentHashMapDynamicGraphSpec extends WordSpec with Matchers {
+  "An Efficient SynchronizedDynamicDirectedGraphSpec" should {
+    "support adding nodes" in {
+      val graph = new ConcurrentHashMapDynamicGraph()
+      for (i <- 0 until 3) {
+        graph.nodeCount shouldEqual i
+        graph.edgeCount shouldEqual 0
+        graph.getOrCreateNode(10 * i) // non-contiguous
+      }
+      graph.getOrCreateNode(10) // Accessing again should increase node count
+      graph.nodeCount shouldEqual 3
+      graph.existsNodeId(1000000) shouldEqual false
+    }
+
+    "support adding edges" in {
+      val graph = new ConcurrentHashMapDynamicGraph()
+      graph.addEdge(1, 2)
+      // For now, addEdge allows duplicates.  graph.addEdge(1, 2) // Test duplicate elimination
+      graph.edgeCount shouldEqual 1
+      val node1 = graph.getNodeById(1).get
+      node1.inboundNodes.toList shouldEqual ( List())
+      node1.outboundNodes.toList shouldEqual (List(2))
+      val node2 = graph.getNodeById(2).get
+      node2.inboundNodes.toList shouldEqual (List(1))
+      node2.outboundNodes.toList shouldEqual (List())
+
+      // test immutability of outboundNodes
+      val oldOutboundNodes1: Seq[Int] = node1.outboundNodes()
+      val oldInboundNodes2: Seq[Int] = node2.inboundNodes()
+      graph.addEdge(1, 10)
+      graph.addEdge(200, 2)
+      oldOutboundNodes1.toList shouldEqual (List(2))
+      oldInboundNodes2.toList shouldEqual (List(1))
+
+      // Test multi-edge
+      graph.addEdge(1, 2)
+      node1.inboundNodes.toList shouldEqual (List())
+      node1.outboundNodes.toList shouldEqual (List(2, 10, 2))
+      node2.inboundNodes.toList shouldEqual (List(1, 200, 1))
+      node2.outboundNodes.toList shouldEqual (List())
+    }
+
+    "support concurrent writing" in {
```

Comment:
I remember starting to use futures for this based on SynchronizedDynamicGraphSpec, but there was some reason they were either more awkward than threads, or I felt it was harder to actually get concurrent execution.  I think the threaded code is fairly clear, given the intent is concurrent execution and not computing a result.